core/src/main/scala/pureconfig/ConfigSource.scala

```diff
@@ -0,0 +1,320 @@
+package pureconfig
+
+import java.io.File
+import java.net.URL
+import java.nio.file.Path
+
+import scala.reflect.ClassTag
+
+import com.typesafe.config._
+import pureconfig.ConfigReader.Result
+import pureconfig.backend.{ ConfigFactoryWrapper, PathUtil }
+import pureconfig.error.{ CannotRead, ConfigReaderException, ConfigReaderFailures }
+
+/**
+ * A representation of a source from which `ConfigValue`s can be loaded, such as a file or a URL.
+ *
+ * A source allows users to load configs from this source as any type for which a `ConfigReader` is
+ * available. Raw configs can also be retrieved as a `ConfigValue`, a `ConfigCursor` or a
+ * `FluentConfigCursor`. Before using any of the loading methods described, Users can opt to focus
+ * on a specific part of a config by specifying a namespace.
+ *
+ * All config loading methods are lazy and defer resolution of references until needed.
+ */
+trait ConfigSource {
+
+  /**
+   * Retrieves a `ConfigValue` from this source.
+   *
+   * @return a `ConfigValue` retrieved from this source.
+   */
+  def value(): Result[ConfigValue]
+
+  /**
+   * Returns a cursor for a `ConfigValue` retrieved from this source.
+   *
+   * @return a cursor for a `ConfigValue` retrieved from this source.
+   */
+  def cursor(): Result[ConfigCursor] =
+    value().right.map(ConfigCursor(_, Nil))
+
+  /**
+   * Returns a fluent cursor for a `ConfigValue` retrieved from this source.
+   *
+   * @return a fluent cursor for a `ConfigValue` retrieved from this source.
+   */
+  def fluentCursor(): FluentConfigCursor =
+    FluentConfigCursor(cursor())
+
+  /**
+   * Navigates through the config to focus on a namespace.
+   *
+   * @param namespace the namespace to focus on
+   * @return a new `ConfigSource` focused on the given namespace.
+   */
+  def at(namespace: String): ConfigSource =
+    ConfigSource.fromCursor(fluentCursor().at(PathUtil.splitPath(namespace).map(p => p: PathSegment): _*))
+
+  /**
+   * Loads a configuration of type `A` from this source.
+   *
+   * @tparam A the type of the config to be loaded
+   * @return A `Right` with the configuration if it is possible to create an instance of type
+   *         `A` from this source, a `Failure` with details on why it isn't possible otherwise
+   */
+  def load[A](implicit reader: Derivation[ConfigReader[A]]): Result[A] =
```

Comment:
It occurs to me that about the only thing I don't like about this new approach is the aesthetics feel a little funny. (Feel free to file the rest of this comment in the appropriate rubbish bin for feelings from folks who don't really contribute.)

I think I'd rather see `load` and `loadOrThrow` move to the `package object pureconfig` and take a `ConfigSource` as a argument.

`load[Config](ConfigSource.applicationConf(ConfigSource.file(path)))`
_feels_ nicer than
`ConfigSource.applicationConf(ConfigSource.file(path)).load[Config]`

I'd also deprecate `pureconfig.loadConfig` in favor of `load(ConfigSource.default)`. 

Feel free to ignore this.

Reply:
Your comments are always welcome and they are contributions as good as any other :)

As a fan of fluent interfaces, I actually prefer the readability of chaining `load` after building a source (skipping the extra set of parentheses and keeping the sequence of operations readable left-to-right, top-to-bottom). I'm not too strong about it though - and there's also the possibility of defining both approaches, but I think that it would be one case where having multiple ways of doing the same thing might be confusing. What do you think? What's your preference, @jcazevedo?

I have a somewhat stronger opinion against having static methods or classes with very general names. `pureconfig` is typically imported as `import pureconfig._`; having a method called simply `load` can easily cause name clashes with other modules, besides the fact that `load` is not a descriptive name at all for client code (as opposed to `loadConfig`, that clearly indicates what we are loading).

Reply:
> As a fan of fluent interfaces, I actually prefer the readability of chaining `load` after building a source (skipping the extra set of parentheses and keeping the sequence of operations readable left-to-right, top-to-bottom). I'm not too strong about it though - and there's also the possibility of defining both approaches, but I think that it would be one case where having multiple ways of doing the same thing might be confusing. What do you think? What's your preference, @jcazevedo?

I prefer the fluent interface. 

With the change proposed by this pull request, I also wouldn't mind deprecating the no arguments versions of `pureconfig.load` and `pureconfig.loadConfigOrThrow` and keep the "entrypoint" methods in the `ConfigSource` interface only. Since part of the motivation of this pull request was to avoid the explosion of methods to load configurations, I think it'd simplify things if we didn't have those methods in the `pureconfig` package.

Reply:
I left them there because my guess is that they are the main and only entrypoints for the great majority of our users and `loadConfig[A]` feels much simpler and to the point than `ConfigSource.default.load[A]`. Like the design around config resolution, I feel we should design to give as much flexibility as possible to power users, but I'd still like to spare casual users of the nitty-gritty details - chances are that most of them will never need to know about `ConfigSource` or most of the other awesome PureConfig features. If we keep to the rule of keeping only no-arg `loadConfig` and `loadConfigOrThrow` in the object, there won't be any explosion of methods.

That said, I'm not too strong about these two as with the other design decisions. I'll happily deprecate `load` and `loadConfigOrThrow` if you think it will simplify the API 🙂 @jcazevedo, @leifwickland, do you still prefer them to be deprecated?

Reply:
I'd lean toward deprecating the no-argument `loadConfig` so that there aren't two substantially different ways to load a config.

Reply:
I also lean towards deprecating them and only recommend using the `ConfigSource` way to load a config.

Reply:
Got it 👍 Done in 0fac76a